---
title: "Agentic AI for Adaptive Pharmacogenomic Biomarker Discovery in Cancer"
output: pdf_document
---

```{r load_packages, eval=TRUE, echo=FALSE, include=FALSE}
suppressPackageStartupMessages({
  library(PharmacoGx)
  library(tidyverse)
  library(genefu)
  library(here)
})
```

## Applications in Preclinical Biomarker Discovery {-}

Pharmacogenomic datasets provide a wealth of biological information across various
experiments and sequencing technologies for a large collection of preclinical samples.
While this makes for an ideal data source for preliminary biomarker discovery, 
the data container structure can be difficult to navigate.
An agentic framework, provided with tools from the PharmacoGx package, can 
expedite bioinformatic analysis.

### Case scenario 1: 
Gene expression-based subtypes of breast cancer are widely studied to be associated
with response to treatment. 

The following dataset was downloaded from [orcestra.ca](https://orcestra.ca). The
dataset is stored as a `PharmacoSet` object from the PharmacoGx package. The dataset
contains 74 breast cancer samples with 9756 drug response experiments.

```{r}
gray <- readRDS(here("./data/rawdata/PSet_GRAY2017.rds")) |> updateObject()
gray
```

*Task: compute pam50 subtypes*

```{r}
# load in genefu PAM50 subtyping model
data(pam50.robust)

# get gene expression
gray_rna <- summarizeMolecularProfiles(
  gray, 
  mDataType = "Kallisto_0.46.1.rnaseq"
) |> suppressMessages()
rna <- t(assay(gray_rna)) |> as.data.frame()

# create metadata file
meta <- gray_rna@elementMetadata[,c("gene_id", "gene_name")] |> as.data.frame()
colnames(meta) <- c("Ensembl", "Gene.Symbol")
colnames(rna) <- meta$Gene.Symbol[match(colnames(rna), meta$Ensembl)]

# get subtype predictions for PAM50
pam50 <- molecular.subtyping(
  sbt.model = "pam50", 
  data = rna,
  annot = meta, 
  do.mapping = FALSE
)
gray_pam50 <- cbind(
  Subtype = as.character(pam50$subtype), 
  as.data.frame(pam50$subtype.proba)
)
head(gray_pam50)
```

*Task: assess if response to HER2-targeted therapies is more prevalent in HER2 subtypes*

```{r}
# get drug sensitivity matrix
gray_sen <- summarizeSensitivityProfiles(
    gray,
    sensitivity.measure = "aac_recomputed",
    summary.stat = "median",
    verbose = TRUE
) |> t() |> as.data.frame()

# add drugs of interest
toPlot <- cbind(gray_pam50, gray_sen[,c("Trastuzumab", "Lapatinib")])
her2_toPlot <- toPlot %>%
  dplyr::select(Subtype, Trastuzumab, Lapatinib) %>%
  pivot_longer(
    cols = c(Trastuzumab, Lapatinib),
    names_to = "Drug",
    values_to = "AAC"
  )
her2_toPlot$label <- ifelse(her2_toPlot$Subtype == "Her2", "Her2", "Other")

ggplot(her2_toPlot, aes(x = Subtype, y = AAC, fill = label)) +
  geom_boxplot() + facet_wrap(~Drug) +
  theme_minimal() + theme(panel.border = element_rect()) +
  scale_fill_manual("Subtype", values = c("#5E4B56", "#A690A4"))
```


### Case scenario 2:
High tumour cellularity can be associated with more aggressive cancers. 
Gene signatures that are surrogate markers of tumour cellularity may be predictive 
of response to immunotherapies.

The code retrieves a dataset from [orcestra.ca](https://orcestra.ca). The
dataset is stored as a `PharmacoSet` object from the PharmacoGx package. The dataset
contains 788 samples 675 cell lines with mutation and transcriptomic profiles, along 
with 16,688 drug response experiments.

```{r}
gcsi <- readRDS(here("./data/rawdata/gCSI.rds")) |> updateObject()
gcsi
```
. . .

*Task: extract the gene expression and mutation profiles*

```{r}
gcsi_rna <- summarizeMolecularProfiles(
  gcsi, 
  mDataType = "Kallisto_0.46.1.rnaseq"
) |> suppressMessages()
gcsi_mut <- summarizeMolecularProfiles(
  gcsi,
  mDataType = "mutation",
  summary.stat = "and"
) |> suppressMessages()
```

*Task: get treatment response data*

```{r}
gcsi_sen <- summarizeSensitivityProfiles(
    gcsi,
    sensitivity.measure = "aac_recomputed",
    summary.stat = "median",
    verbose = TRUE
)
```


*Task: Find samples with lapatinib*

```{r}
egfr_tki <- c("Gefitinib", "Erlotinib", "Lapatinib")
genes <- c("EGFR", "BRCA1", "BRCA2")

# subset drug sensitivity
sen <- t(gcsi_sen[egfr_tki,]) |> as.data.frame()

# subset mutations
mut <- t(assay(gcsi_mut)) |> as.data.frame()
mut <- mut[,genes]

# subset gene expression
rna <- assay(gcsi_rna) |> as.data.frame()
keep <- gcsi_rna@elementMetadata$gene_id[gcsi_rna@elementMetadata$gene_name %in% genes]
rna <- t(rna[keep,]) |> as.data.frame()
colnames(rna) <- gcsi_rna@elementMetadata$gene_name[match(colnames(rna), gcsi_rna@elementMetadata$gene_id)]

# keep common cell lines
common <- intersect(rownames(sen), intersect(rownames(rna), rownames(mut)))
sen <- sen[match(common, rownames(sen)),]
rna <- rna[match(common, rownames(rna)),]
mut <- mut[match(common, rownames(mut)),]
```

*Task: is there an association between EGFR expression and TKi response*

```{r}
# plot associations between genes of interest and TKi
toPlot <- data.frame(
  sample = common,
  EGFR = rna[["EGFR"]],
  Lapatinib = sen[["Lapatinib"]],
  Erlotinib = sen[["Erlotinib"]],
  Gefitinib = sen[["Gefitinib"]]
)

# plot associations between genes of interest and TKi
plot_egfr <- function(drug) {
  return(
    ggplot(toPlot, aes(x = EGFR, y = .data[[drug]])) +
      geom_point() + geom_smooth(method = "lm") +
      theme_minimal() + 
      labs(x = "EGFR expression (TPM)", y = paste(drug, "response (AAC)"), title = drug)
  )
} 

p1 <- plot_egfr("Lapatinib")
p2 <- plot_egfr("Erlotinib")
p3 <- plot_egfr("Gefitinib")

ggarrange(p1, p2, p3, ncol = 3)
```

*Task: does BRCA mutation status influence the strength of association*

```{r}
# get samples with BRCA1 mutation
mutated <- mut[!is.na(mut$BRCA1),] |> rownames()
toPlot$brca1 <- ifelse(toPlot$sample %in% mutated, "Mutated", "Unmutated")

# plot association faceted by BRCA1 mutation status
plot_egfr <- function(drug) {
  return(
    ggplot(toPlot, aes(x = EGFR, y = .data[[drug]])) +
      geom_point() + geom_smooth(method = "lm") +
      facet_wrap(~brca1, ncol = 1) +
      theme_minimal() + 
      labs(x = "EGFR expression (TPM)", y = paste(drug, "response (AAC)"), title = drug)
  )
} 

p1 <- plot_egfr("Lapatinib")
p2 <- plot_egfr("Erlotinib")
p3 <- plot_egfr("Gefitinib")

ggarrange(p1, p2, p3, ncol = 3)
```
